xla/service/cpu/onednn_contraction_rewriter.cc

```diff
@@ -612,141 +644,125 @@ class OneDnnContractionRewriteVisitor : public DfsHloRewriteVisitor {
-
-    // TODO(intel-tf): Remove this restriction once oneDNN has an optimized
-    // implementation for broadcasted add across all dimensions.
-    OneDnnFusionConfig_FusionKind kind = OneDnnFusionConfig::UNDEFINED;
-    kind =
-        (addend->shape().rank() == 1)
-            ? (GetKernelConfig<config>(&backend_config)->fusions().ops().empty()
-                   ? OneDnnFusionConfig::BIAS
-                   : OneDnnFusionConfig::UNDEFINED)
-            : OneDnnFusionConfig::BINARY_ADD;
-    if (kind == OneDnnFusionConfig::UNDEFINED) return absl::OkStatus();
-
-    GetKernelConfig<config>(&backend_config)->mutable_fusions()->add_ops(kind);
-
-    if (optional_addend_broadcast) {
-      GetKernelConfig<config>(&backend_config)
-          ->mutable_optimization_config()
-          ->set_bias_broadcast(true);
-    }
-    TF_RETURN_IF_ERROR(custom_call->set_backend_config(*backend_config));
-
-    HloInstruction* new_instr;
-    // If matched pattern has custom-call -> bitcast -> add, then we need to
-    // insert bitcast after the new fusion to maintain the correct shape
-    // (new-custom-call -> bitcast). Also, this will optionally be followed
-    // by -> convert for bf16 case to avoid datatype mismatch.
-    if (optional_contraction_bitcast != nullptr &&
-        optional_contraction_bitcast->opcode() == HloOpcode::kBitcast) {
-      if (optional_contraction_convert != nullptr &&
-          optional_contraction_convert->opcode() == HloOpcode::kConvert) {
-        auto bitcast_call =
-            custom_call->AddInstruction(HloInstruction::CreateBitcast(
-                ShapeUtil::ChangeElementType(
-                    instr->shape(), custom_call->shape().element_type()),
-                custom_call));
-        new_instr = bitcast_call->AddInstruction(HloInstruction::CreateConvert(
-            ShapeUtil::ChangeElementType(
-                bitcast_call->shape(),
-                optional_contraction_convert->shape().element_type()),
-            bitcast_call));
-      } else {
-        new_instr = custom_call->AddInstruction(
-            HloInstruction::CreateBitcast(instr->shape(), custom_call));
+      auto custom_call = Cast<HloCustomCallInstruction>(
+          instr->AddInstruction(contraction->CloneWithNewOperands(
+              contraction->shape(), new_operands)));
+
+      auto backend_config = custom_call->backend_config<BackendConfig>();
+      auto fusions_config = GetFusionsConfig(&backend_config);
+      auto optimization_config = GetOptimizationsConfig(&backend_config);
+      // TODO(intel-tf): Remove this restriction once oneDNN has an optimized
+      // implementation for broadcasted add across all dimensions.
+      OneDnnFusionConfig_FusionKind kind = OneDnnFusionConfig::UNDEFINED;
+      kind =
+          (addend->shape().rank() == 1)
+              ? (fusions_config->ops().empty() ? OneDnnFusionConfig::BIAS
+                                               : OneDnnFusionConfig::UNDEFINED)
```

Comment:
Does this mean the code doesn't support having 1D addend (bias) and fused ops at the same time? Please document this as a comment in the code.

Reply:
This should ensure that 1D addends are fused only when they are the first post-op operation. So the following should be acceptable:

1. Bias (1D) + <0 or more non-add post-op>
2. Bias (1D) + <0 or more non-add post-op> + Add (non-1D) + <0 or more non-add post-op>  

Following is not allowed:

1. Bias (1D) + <0 or more non-add post-op> + Add (1D) + <0 or more non-add post-op> 

This condition was added a few months ago because at the time oneDNN had optimized implementations for broadcasted add operations across certain dimensions only. As a result, some cases defaulted to the ref implementation, which significantly impacted performance.
We can re-evaluate this restriction with the latest oneDNN release and/or relax this a bit by blocking only those 1D cases where broadcasting occurs along some specific dimensions.